src/components/Editor/EditorInput/EditorInput.js

```diff
@@ -0,0 +1,110 @@
+import React, { createRef, useEffect, useRef, useState } from 'react'
+import { DragDropContext } from '@hello-pangea/dnd'
+import { useDispatch, useSelector } from 'react-redux'
+import { closeFile, setFocussedFileIndex, setOpenFiles } from '../EditorSlice'
+import { TabPanel, Tabs } from 'react-tabs'
+import Button from '../../Button/Button'
+import { CloseIcon } from '../../../Icons'
+import EditorPanel from '../EditorPanel/EditorPanel'
+import DraggableTab from '../DraggableTabs/DraggableTab'
+import DroppableTabList from '../DraggableTabs/DroppableTabList'
+import RunBar from '../../RunButton/RunBar'
+
+import './EditorInput.scss'
+
+const EditorInput = () => {
+  const project = useSelector((state) => state.editor.project)
+  const openFiles = useSelector((state) => state.editor.openFiles)
+  const focussedFileIndices = useSelector((state) => state.editor.focussedFileIndices)
+  const dispatch = useDispatch()
+
+  const onDragStart = (input) => {
+    const { source } = input
+    dispatch(setFocussedFileIndex({panelIndex: parseInt(source.droppableId), fileIndex: source.index}))
+  }
+
+  const onDragEnd = (result) => {
+    if(!result.destination) return
+    const { source, destination } = result
+    var openFilesData = [...openFiles]
+
+    var oldPane = [...openFilesData[source.droppableId]]
+    const [removed] = oldPane.splice(source.index, 1)
+    openFilesData[source.droppableId] = [...oldPane]
+
+    var newPane = [...openFilesData[destination.droppableId]]
+    newPane.splice(destination.index, 0, removed)
+    openFilesData[destination.droppableId] = [...newPane]
+    dispatch(setOpenFiles(openFilesData))
+    dispatch(setFocussedFileIndex({panelIndex: parseInt(destination.droppableId), fileIndex: destination.index}))
+    if (destination.droppableId !== source.droppableId) {
+      dispatch(setFocussedFileIndex({panelIndex: parseInt(source.droppableId), fileIndex: Math.max(source.index - 1, 0)}))
+    }
+  }
+
+  const closeFileTab = (e, fileName) => {
+    e.stopPropagation()
+    dispatch(closeFile(fileName))
+  }
+
+  const [numberOfComponents, setNumberOfComponents] = useState(project.components.length)
+  let tabRefs = useRef(project.components.map(createRef))
+
+  useEffect(() => {
+    setNumberOfComponents(project.components.length)
+    Array(project.components.length).fill().forEach((_, i) => {
```

Comment:
Might be missing something but do you need the call to `fill()` here? I must admit I agonised over this a bit as I'd have just used an old fashioned `for` loop I think, but this works :)

Reply:
Right, I think I'd have been compelled to use a simple `for` loop `for (let i = 0; i < project.components.length; i++) {`, but this works :)